File: src/test/java/com/uapp_llc/test/comparator/NotNullComparator.java
package com.uapp_llc.test.comparator;

import java.util.Comparator;

public final class NotNullComparator<T> implements Comparator<T> {

  private static final NotNullComparator<?> LEFT = new NotNullComparator<>(true, false);
  private static final NotNullComparator<?> RIGHT = new NotNullComparator<>(false, true);
  private static final NotNullComparator<?> BOTH = new NotNullComparator<>(true, true);

  private final boolean leftNotNull;
  private final boolean rightNotNull;

  private NotNullComparator(boolean leftNotNull, boolean rightNotNull) {
    this.leftNotNull = leftNotNull;
    this.rightNotNull = rightNotNull;
  }

  public static <T> NotNullComparator<T> leftNotNull() {
    return (NotNullComparator<T>) LEFT;
  }

  public static <T> NotNullComparator<T> rightNotNull() {
    return (NotNullComparator<T>) RIGHT;
  }

  public static <T> NotNullComparator<T> bothNotNull() {
    return (NotNullComparator<T>) BOTH;
  }

  @Override
  public int compare(T left, T right) {
    boolean result;

    if (leftNotNull && rightNotNull) {
      result = (left != null) && (right != null);
    } else if (leftNotNull) {
      result = left != null;
    } else {
      result = right != null;
    }

    return result ? 0 : 1;
  }

}

File: src/test/kotlin/com/uapp_llc/test/comparator/NotNullComparator.kt
package com.uapp_llc.test.comparator

import java.util.Comparator

class NotNullComparator<T> private constructor(

    private val leftNotNull: Boolean,
    private val rightNotNull: Boolean

) : Comparator<T> {

  override fun compare(left: T, right: T): Int {
    val result: Boolean =
        if (leftNotNull && rightNotNull) {
          (left != null) && (right != null)
        } else if (leftNotNull) {
          left != null
        } else {
          right != null
        }

    return if (result) 0 else 1
  }


  companion object {

    private val left: NotNullComparator<Any> =
        NotNullComparator(leftNotNull = true, rightNotNull = false)
    private val right: NotNullComparator<Any> =
        NotNullComparator(leftNotNull = false, rightNotNull = true)
    private val both: NotNullComparator<Any> =
        NotNullComparator(leftNotNull = true, rightNotNull = true)

    @JvmStatic
    fun <T> leftNotNull(): NotNullComparator<in T> {
      return left as NotNullComparator<in T>
    }

    @JvmStatic
    fun <T> rightNotNull(): NotNullComparator<in T> {
      return right as NotNullComparator<in T>
    }

    @JvmStatic
    fun <T> bothNotNull(): NotNullComparator<in T> {
      return both as NotNullComparator<in T>
    }

  }

}
